Lib/Src/CUartCom.cpp

```diff
@@ -10,67 +10,79 @@
 
 #include "CUartCom.h"
 
+CUartCom *CUartCom::sp_UART_1 = NULL;
+CUartCom *CUartCom::sp_UART_2 = NULL;
+
 /**
  * @brief Construct UART communication object.
- *
  * @param p_huart Pointer to UART hardware control register structure.
+ * @param name Name of the instance
+ * @note If no USART_DE Pin is needed, the BREATHING Pin will turn on when UART
+ * is sending data
  */
-CUartCom::CUartCom(UART_HandleTypeDef *p_huart) : mp_huart(p_huart)
+CUartCom::CUartCom(UART_HandleTypeDef *p_huart, const std::string name)
+    : IComChannel(name),
+      mp_huart(p_huart),
+      m_uart_de_pin(BREATHING_GPIO_Port, BREATHING_Pin)
 {
     // Check mp_huart is not null
     if (!mp_huart)
     {
-        // Hang application with a toggling ALARM Pin
-        while (1)
-        {
-            // TODO: cannot use alarm pin. No space for it due to one less pin
-            // available on this MCU. HAL_GPIO_TogglePin(ALARM_GPIO_Port,
-            // ALARM_Pin);
-            HAL_Delay(200);
-        }
+        Error_Handler();
     }
 }
 
 /**
  * @brief Construct UART communication object with half-duplex mode.
- *
  * @param p_huart Pointer to UART hardware control register structure.
+ * @param name Name of the CUartCom Instance
  * @param uart_de_port Pointer to GPIO port.
  * @param uart_de_pin GPIO pin.
  */
 CUartCom::CUartCom(UART_HandleTypeDef *p_huart,
+                   const std::string name,
                    GPIO_TypeDef *uart_de_port,
                    uint16_t uart_de_pin)
-    : mp_huart(p_huart),
-      m_uart_de_port(uart_de_port),
-      m_uart_de_pin(uart_de_pin)
+    : IComChannel(name),
+      mp_huart(p_huart),
+      m_uart_de_pin(uart_de_port, uart_de_pin)
 {
     // Check mp_huart is not null
     if (!mp_huart)
     {
-        // Hang application with a toggling ALARM Pin
-        while (1)
-        {
-            // TODO: cannot hardcode stuff like this into classes.
-            //            HAL_GPIO_TogglePin(ALARM_GPIO_Port, ALARM_Pin);
-            HAL_Delay(200);
-        }
+        Error_Handler();
+    }
+}
+
+/**
+ * @brief Enables UART reception with interrupts
+ * @note Interrupts must be enabled for the UART port in order to use this
+ * function
+ *
+ */
+void CUartCom::startReception()
+{
+    HAL_UART_Receive_IT(mp_huart, &m_rx_char, 1);
+    if (mp_huart->Instance == USART1)
+    {
+        sp_UART_1 = this;
```

Comment:
Ask yourself: does `sp_uart_1` mean that this static pointer belongs to the first instance of `CUartCom` class, or does it mean it belongs to whatever instance is in control of UART1? How does any specific instance of the class know which hardware resource belongs to it? What happens when we want to use UART2, UART3 and UART7 instead?

Reply:
This is possibly my mistake not fully understanding your code. I am not going to delete the comment yet, but don't rush correcting anything.

Reply:
OK. I am pretty sure that what you've done as far as associating numbers of the hardware instances with numbers of the static variables is fine. It's the other stuff I am not so happy about. We can close this comment when others are resolved.